apollo-router/src/services/router_service.rs

```diff
@@ -339,61 +324,208 @@ impl RouterService {
+                                        MULTIPART_DEFER_CONTENT_TYPE
+                                    ))
+                                    .extension_code("INVALID_ACCEPT_HEADER")
+                                    .build(),
+                            )
+                            .status_code(StatusCode::NOT_ACCEPTABLE)
+                            .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
+                            .context(context)
+                            .build()?);
                 }
             }
         }
+        Ok(())
+    }
+
+    async fn call_inner(&self, req: RouterRequest) -> Result<RouterResponse, BoxError> {
+        let context = req.context.clone();
+
+        let supergraph_requests = match self.translate_request(req).await {
+            Ok(requests) => requests,
+            Err(err) => {
+                ::tracing::error!(
+                    monotonic_counter.apollo_router_http_requests_total = 1u64,
+                    status = %err.status.as_u16(),
+                    error = %err.error,
+                );
+
+                return router::Response::error_builder()
+                    .error(
+                        graphql::Error::builder()
+                            .message(String::from("Invalid GraphQL request"))
+                            .extension_code(err.extension_code)
+                            .extension("details", err.extension_details)
+                            .build(),
+                    )
+                    .status_code(err.status)
+                    .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
+                    .context(context)
+                    .build();
+            }
+        };
+
+        let mut results = vec![];
+
+        for supergraph_request in supergraph_requests {
+            self.process_supergraph_request(&mut results, supergraph_request)
+                .await?;
+        }
+        // If we only have one result, go ahead and return it. Otherwise, create a new result
+        // which is an array of all results.
+        if results.len() == 1 {
+            Ok(results.pop().expect("should be at least one response"))
+        } else {
+            let first = results.pop().expect("we should have at least one response");
+            let (parts, body) = first.response.into_parts();
+            let context = first.context;
+            let mut array_result = "[".to_string();
+            array_result +=
+                &String::from_utf8(hyper::body::to_bytes(body).await?.to_vec()).unwrap();
+            for result in results {
+                array_result += ", ";
+                array_result += &String::from_utf8(
+                    hyper::body::to_bytes(result.response.into_body())
+                        .await?
+                        .to_vec(),
+                )
+                .unwrap();
+            }
+            array_result += "]";
```

Comment:
I'm wondering if it would not be more optimized to only use bytes and then convert it to string if needed afterwards. For example we used [this](https://github.com/apollographql/router/blob/dev/apollo-router/src/protocols/multipart.rs#L149) and [this](https://github.com/apollographql/router/blob/dev/apollo-router/src/protocols/multipart.rs#L140) in the multipart protocol